### PR TITLE
Feature: Add OpenID Connect SSO support via django-allauth

### PR DESCRIPTION
**Pull request by smkent**
## Proposed change

I use OpenID Connect (OIDC) for SSO with my self-hosted web apps. I've been excited to try out paperless-ngx for a while, but there is currently no support for OpenID Connect. This change adds support for OpenID Connect for SSO via [django-allauth](https://github.com/pennersr/django-allauth), with the potential for also supporting [django-allauth's numerous SSO options](https://django-allauth.readthedocs.io/en/latest/providers.html).

OpenID Connect support enables SSO use via self-hosted providers like [Authentik](https://github.com/goauthentik/authentik) and [Keycloak](https://github.com/keycloak/keycloak).

~**Note**: This PR is based on [my upstream PR in django-allauth](https://github.com/pennersr/django-allauth/pull/3165). This code works with a small Dockerfile change (see below) but should be considered a working concept until my upstream PR is merged.~ [django-allauth 0.52.0 has been released which contains support for OpenID Connect!](https://github.com/pennersr/django-allauth/blob/0.52.0/ChangeLog.rst)

Sample OIDC configuration:

```sh
PAPERLESS_SSO_OIDC_NAME="Test OIDC"
PAPERLESS_SSO_OIDC_URL="https://some.oidc.server.example.com"
PAPERLESS_SSO_OIDC_CLIENT_ID="some-client-id-from-your-oidc-server"
PAPERLESS_SSO_OIDC_SECRET="accompanying-secret-from-your-oidc-server"
```

With no SSO configured, this PR does not change the login behavior:

paperless-password

However, SSO options will now appear on the login screen when configured:

paperless-both

If SSO is configured, it is also possible to hide the password form via a new setting `PAPERLESS_LOGIN_HIDE_PASSWORD_FORM`:

paperless-sso

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other (please explain)

## Checklist:

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [x] If applicable, I have tested my code for new features & regressions on both ✅ mobile & ✅ desktop devices, using the latest version of major browsers.
- [ ] If applicable, I have checked that all tests pass, see [documentation](https://paperless-ngx.readthedocs.io/en/latest/extending.html#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://paperless-ngx.readthedocs.io/en/latest/extending.html#code-formatting-with-pre-commit-hooks).
- [ ] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

### CODE REVIEW

**Comment by shamoon** [comment]
This is awesome, thanks for the PR. Let's keep an eye on your upstream PR and can circle back to this hopefully soon. When the time comes I think we should update the documentation to reflect this too. Bravo 👏

**Comment by olant** [comment]
Looking forward for this PR being approved!

**Comment by ciphermenial** [comment]
Not sure if this will be useful at all but I did something similar a while ago and did a write up about it. https://ciphermenial.github.io/posts/adding-oauth-to-paperless/

**Comment by MartinVerges** [comment]
please merge! Using a central auth is an essential feature in my opinion

**Comment by stumpylog** [comment]
This PR is a draft for the reasons stated when it opened.  Once django-allauth has merged and released, it will get reviewed and merged.

**Comment by shamoon** [comment]
well well https://github.com/pennersr/django-allauth/pull/3165 !

**Comment by Celant** [comment]
Alas just waiting on the maintainer of that package to push a new release first, but yes, good news regardless!

**Comment by hendrik1120** [comment]
Version 0.52.0 of django-allauth has just been released. Can't wait to test this with authelia.

**Comment by ikaruswill** [comment]
I've tested Django allauth over on Tandoor recipes, works pretty well Authelia!

**Comment by stumpylog** [comment]
@smkent Any free time to resolve the conflicts and mark this as ready?  I think people would love to see it in

**Comment by shamoon** [comment]
I am wondering about the order of things with this vs #2147 , which I know is a beast but would like to get to soon. My thinking was it would be even better to that in before this, but not like that’s a hard requirement or anything

**Comment by smkent** [comment]
I've updated this PR with django-allauth 0.52.0 and merged in `dev` to resolve the merge conflicts.

I'm interested in writing unit tests for this feature. Please let me know what else I can do to help!

**Comment by stumpylog** [comment]
I have no idea how to write testing for the feature, but perhaps the allauth tests can be looked to for an example.  It would be best to have the feature covered.

**Comment by SpiderD555** [comment]
Recently [Gitea had implemented some oauth2 tests](https://github.com/go-gitea/gitea/pull/21390/commits/aa2b20a4867610787eafa881e41b77724db65d18) which can give some idea on the proper flow of tests and which endpoints to call. While oidc is not entirely oauth2, I think it can at least give some general hints about what to test.

File: src/paperless/settings.py
Comment (stumpylog) [inline src/paperless/settings.py:51]: I'm not loving how many unchanged lines have changed in this diff. It appears a bunch of commas were added. It makes it difficult to review. Could you please remove these changes?

**Comment by stumpylog** [comment]
I'd like to start testing the image out and see how it works, but one thing still lacking from the PR is documentation of the feature. 
There's a number of new settings to document and I think one or two setup guides would be very helpful.  These could be generic, or could be a little more specific to some of the common tools like Authelia and Authentik

**Comment by hendrik1120** [comment]
Hi, I already did some basic testing with authelia last week and the setup was very easy except one issue I encountered.
Just adding a new OIDC client in authelia and setting 4 envs went pretty smoothly, but authelia also requires a ```redirect_uri``` to be set on both sides, which must be exactly the same.

Authentication always failed because django-allauth set the ```redirect_uri``` to ```https://``` which I wasn't even using. I couldn't find anything in the documentation to change this, and also manually setting the ```PAPERLESS_URL``` to ```http://``` didn't help.  So I set up a local reverse proxy to not bother with django ssl certs and finally got this working.

Hope this helps if someone is encountering the same issue, and pls let me know if there is a way to change this config.

**Comment by mwllgr** [comment]
Is there a way to test this/does a Docker image already exist? I just switched all my services to Keycloak and can't wait to try out paperless-ngx! :)

**Comment by lazyzyf** [comment]
> redirect_uri



> Hi, I already did some basic testing with authelia last week and the setup was very easy except one issue I encountered. Just adding a new OIDC client in authelia and setting 4 envs went pretty smoothly, but authelia also requires a `redirect_uri` to be set on both sides, which must be exactly the same.
> 
> Authentication always failed because django-allauth set the `redirect_uri` to `https://` which I wasn't even using. I couldn't find anything in the documentation to change this, and also manually setting the `PAPERLESS_URL` to `http://` didn't help. So I set up a local reverse proxy to not bother with django ssl certs and finally got this working.
> 
> Hope this helps if someone is encountering the same issue, and pls let me know if there is a way to change this config.

how do you set the `redirect_uri` in Authelia?

**Comment by lazyzyf** [comment]
> I've tested Django allauth over on Tandoor recipes, works pretty well Authelia!

how did you set the `redirect_uris` in Authelia?

**Comment by stumpylog** [comment]
There's also the pre-built image `ghcr.io/smkent/paperless-ngx:feature-oidc-allauth`

@DennisGaida That's detailed setup, thanks.  I may be able to document setup on this side enough with those.

**Comment by lazyzyf** [comment]
> There's also the pre-built image `ghcr.io/smkent/paperless-ngx:feature-oidc-allauth`
> 
> @DennisGaida That's detailed setup, thanks. I may be able to document setup on this side enough with those.

this image works perfect.

**Review by stumpylog — CHANGES_REQUESTED** [review]
Presumably this is a misconfiguration on my part (but goes to show the need for good documentation on this feature), but I'm not able to get it working with Authelia.  Attempting to log in using OIDC spins for sometime, before timing out.

**Comment by hendrik1120** [comment]
@stumpylog what kind of documentation do you have in mind?
For people with authelia and oidc already set up, it's just as easy as setting 4 variables and using paperless with https.
Only documentation for the variables would be needed and a warning about the redirect uri https issue.

Oidc in general really only makes sense to set up if you have a lot of services to log in to.
Providing a complete setup for keycloak or authelia with all the security considerations will be a long journey. These services also already have their own documentation available on how to integrate common apps, see https://www.authelia.com/integration/openid-connect/. Including paperless in there would also be an option.

**Comment by stumpylog** [comment]
> For people with authelia and oidc already set up, it's just as easy as setting 4 variables and using paperless with https.

Well, I already use Authelia and my own domain for https, but setting the 4 variables merely leads to timeouts.  So I don't find it quite that easy and that's why I feel some additional documentation is still needed, or else there will be lots of reports to deal with.  And clearly, I don't do web stuff often enough to deal with those.

**Comment by jonaswinkler** [comment]
Hello!

First of all, I was able to get it working with Google OIDC. (I don't have Authelia avaiable, and don't really need it.)

Here are a couple issues I noticed:

* Without digging further, authentication with normal users does not work on your branch (i.e., users created via `python manage.py createsuperuser`)
* @shamoon merged the permissions feature into dev, and the user created at the end of the OIDC workflow lacks all permissions to do anything. I've already suggested internally that new users should probably get some minimum default permissions, but here they should *probably* get all admin permissions? Not sure. However, combine that with the issue above and I'm currently unable to do anything in paperless on this branch, since logging in with the other admin users doesn't work.
* There are a lot of settings and code around `PAPERLESS_SSO_MODULES` and `PAPERLESS_SSO_ENABLED`. However, in your opening post you did not mention these at all. Are these two separate use cases? I'm not sure what I would need these for.

**Comment by raelix** [comment]
this would be a great feature!

**Comment by mwllgr** [comment]
Why is this PR marked as _draft_ now?

**Comment by shamoon** [comment]
> Why is this PR marked as _draft_ now?

Because it needs more work. We know folks are eager for the feature to be released, please limit comments to those that are useful to that goal. Thanks all.

**Comment by oblivioncth** [comment]
What [DennisGaida](https://github.com/DennisGaida) noted indirectly touches on this, but something that doesn't seem to have been addressed yet is exactly how this interacts with existing users, and what OIDC claims are used when pulling the user from the provider.

A common flow is to use a user configurable claim (often 'prefered_username' or 'email') that is used when loading the user information from the provider. Them, if that field matches one from an existing user, that user is signed-in, if not, a new user is created with the info from the token or user info endpoint.

Alternatively, some implementations use the 'sub' claim as an ID to tie users in the local database to OIDC users and allow admins to change this value for any given user so that deployments can easily migrate their existing users over to OIDC. An unmatched subject means a new user.

The more flexible this is the better, but at the very least how it's current handled should be documented.

**Comment by mwllgr** [comment]
Hello, I just wanted to share a possibility to use https://github.com/oauth2-proxy/oauth2-proxy for OIDC support while paperless-ngx doesn't support it natively. Only problem I found is that the websocket doesn't want to work, still have to figure that out. This leads to a stuck progress bar on file upload - Might be a problem on the nginx side?

It uses the header authentication mode of paperless-ngx:

`docker-compose.yml`
```yaml
version: "3.4"
services:
  broker:
    image: docker.io/library/redis:7
    restart: unless-stopped
    volumes:
      - /opt/paperless/data/redis:/data

  db:
    image: docker.io/library/postgres:13
    restart: unless-stopped
    volumes:
      - /opt/paperless/data/postgres:/var/lib/postgresql/data
    environment:
      POSTGRES_DB: paperless
      POSTGRES_USER: paperless
      POSTGRES_PASSWORD: paperless

  webserver:
    image: ghcr.io/paperless-ngx/paperless-ngx:latest
    restart: unless-stopped
    depends_on:
      - db
      - broker
      - gotenberg
      - tika
    healthcheck:
      test: ["CMD", "curl", "-fs", "-S", "--max-time", "2", "http://localhost:8000"]
      interval: 30s
      timeout: 10s
      retries: 5
    volumes:
      - /opt/paperless/data/paperless/data:/usr/src/paperless/data
      - /opt/paperless/data/paperless/redis:/usr/src/paperless/media
      - /opt/paperless/data/paperless/export:/usr/src/paperless/export
      - /opt/paperless/data/paperless/consume:/usr/src/paperless/consume
    env_file: docker-compose.env
    environment:
      PAPERLESS_REDIS: redis://broker:6379
      PAPERLESS_DBHOST: db
      PAPERLESS_TIKA_ENABLED: 1
      PAPERLESS_TIKA_GOTENBERG_ENDPOINT: http://gotenberg:3000
      PAPERLESS_TIKA_ENDPOINT: http://tika:9998
      PAPERLESS_ENABLE_HTTP_REMOTE_USER: "true"
      PAPERLESS_HTTP_REMOTE_USER_HEADER_NAME: "HTTP_X_FORWARDED_PREFERRED_USERNAME"
      PAPERLESS_EMAIL_TASK_CRON: "* * * * *"
      PAPERLESS_ENABLE_COMPRESSION: "false"

  gotenberg:
    image: docker.io/gotenberg/gotenberg:7.8
    restart: unless-stopped

    # The gotenberg chromium route is used to convert .eml files. We do not
    # want to allow external content like tracking pixels or even javascript.
    command:
      - "gotenberg"
      - "--chromium-disable-javascript=true"
      - "--chromium-allow-list=file:///tmp/.*"

  tika:
    image: ghcr.io/paperless-ngx/tika:latest
    restart: unless-stopped

  oidc-proxy:
    restart: unless-stopped
    image: quay.io/oauth2-proxy/oauth2-proxy:latest

    environment:
      OAUTH2_PROXY_UPSTREAMS: http://webserver:8000/
      OAUTH2_PROXY_PROVIDER_DISPLAY_NAME: <keycloak-domain>
      OAUTH2_PROXY_PROVIDER: keycloak-oidc
      OAUTH2_PROXY_OIDC_ISSUER_URL: <issuer-url - for keycloak: https://<keycloak-domain>/realms/ic>
      OAUTH2_PROXY_CLIENT_ID: <client-id>
      OAUTH2_PROXY_CLIENT_SECRET: <client-secret>
      OAUTH2_PROXY_REVERSE_PROXY: "true"
      OAUTH2_PROXY_COOKIE_SECURE: "true"
      OAUTH2_PROXY_COOKIE_SECRET: <cookie-secret>
      OAUTH2_PROXY_REDIRECT_URL: https://<paperless-domain>/oauth2/callback
      OAUTH2_PROXY_EMAIL_DOMAINS: "*"
      OAUTH2_PROXY_CODE_CHALLENGE_METHOD: "S256"
      OAUTH2_PROXY_HTTP_ADDRESS: "0.0.0.0:4180"
      OAUTH2_PROXY_SCOPE: "openid email"
      OAUTH2_PROXY_SESSION_COOKIE_MINIMAL: "true"
      OAUTH2_PROXY_OIDC_EXTRA_AUDIENCES: <keycloak-client-id>
      OAUTH2_PROXY_SESSION_STORE_TYPE: "redis"
      OAUTH2_PROXY_REDIS_CONNECTION_URL: "redis://broker"
      OAUTH2_PROXY_SKIP_PROVIDER_BUTTON: "true"
      OAUTH2_PROXY_FOOTER: "-"
      OAUTH2_PROXY_BANNER: "-"
      OAUTH2_PROXY_CUSTOM_SIGN_IN_LOGO: "-"
      OAUTH2_PROXY_PASS_USER_HEADERS: "true"

    ports:
      - 127.0.0.1:8141:4180
 ```

Don't forget to customize `docker-compose.env` with your `PAPERLESS_URL`.

**Comment by dkowis** [comment]
Can confirm, the oauth2 proxy works great, and I just set up a bypass for the API endpoint as well, so that token authentication works fine. I'm using basically this exact setup.

**Comment by XtremeOwnageDotCom** [comment]
Beware- there are multiple vulnerabilities with the current release of oauth2-proxy. Its last release was back in Oct last year. 

Multiple tickets open for it...

**Comment by AndrewNatoli** [comment]
## Using Authentik 

I see the discussion about working with existing accounts and getting more information on the configuration values, so here's my experience using Authentik. 

### Regarding linking existing accounts
In Authentik I created an oAuth2 provider. Under "Advanced protocol settings" there's an option called `Subject Mode`.

The options look like this:
image

When I originally set up PaperlessNgx with the base image, I registered with the same username I use for Authentik, hoping the Proxy auth example from [PaperlessNg](https://goauthentik.io/integrations/services/paperless-ng/) would still work-- it doesn't seem to, so thanks for making this feature!

Now then, I set up my provider to have its `Subject Mode` "Based on the User's username." This means if I was registered as `cooluser` on Authentik, it would tell Paperless my name is `cooluser`. It _does_ do that in the `socialaccount_socialaccount` table under the `uid` column, but it created a _new_ user in `auth_user` with the `username` of `cooluser3` (which seems to be a random number instead of sequential).

I deleted the new user records from `account_emailaddress`, `socialaccount_socialaccount`, and `auth_user` (I believe that's the right order to avoid FK constraints for anyone who tries this themselves) then I switched the `Subject mode` to try email address. 

The email address matched the one on my base Paperless account but still created a new user. 

Ultimately I linked my Authentik login to my existing account by hand. I deleted the `account_emailaddress` and `auth_user` records for the new account, then went to the new record in `socialaccount_socialaccount` and revised the `user_id` to match the ID for my existing account.

I also tried with and without these environment variables which I may have interpreted wrong from the README:

```
      PAPERLESS_SSO_AUTO_LINK: FALSE
      PAPERLESS_SSO_AUTO_LINK_MULTIPLE: FALSE
```

### My environment variables for Authentik

```
      PAPERLESS_SSO_OIDC_NAME: authentik
      PAPERLESS_SSO_OIDC_URL: "https://<my authentik server>/application/o/<provider slug>/"
      PAPERLESS_SSO_OIDC_CLIENT_ID: <client id from provider>
      PAPERLESS_SSO_OIDC_SECRET: <secret from provider>
```
      
 The URL is the value of `OpenID Configuration Issuer` listed on the provider page.
 
 ### Authentik Provider settings
 - OAuth2 Provider
 - Authorization flow-- I'm using default explicit
 - Client type is confidential
 - Redirect URL is `https://<authentik host>/accounts/authentik/login/callback/`
 
 The `authentik` in my redirect URL matches the `PAPERLESS_SSO_OIDC_NAME` value.
 
 --- 
 
Hope this helps with moving this feature forward!

**Comment by shamoon** [comment]
@smkent its been many months with no activity so we’re going to close this. Again, I think others are interested in this feature so we hope to hear from you (or anyone else who is interested in moving this forward) in the future. Thank you for the contribution

**Pull request closed without merging**
